### ISSUE TITLE
dump_transcripts promises comparability it does not deliver, and records none of the context that makes an old transcript usable

### ISSUE THREAD

**Issue by MatthewLacerda2** [MatthewLacerda2/TinyRefinementModel#203]
`instruments/dump_transcripts.py` exists and its docstring already states the goal:

> *"Fixed-prompt transcripts per checkpoint — the systematized vibes eval. Same prompts, same sampling settings, every time: quality becomes comparable across checkpoints instead of living in scrollback."*

It does not deliver that, and it records none of the context that would make an old transcript usable later. This is the class of thing that **cannot be backfilled** — see "Why this is time-sensitive" below.

## 1. The output is not reproducible

`--temperature 0.5` with no seed. `generate_text` draws a fresh key per call, so running the same checkpoint on the same prompts twice gives different text. "Same sampling settings" is not the same as same sample.

The fix is a fixed `PRNGKey`, **not** greedy decoding. Greedy loops more than sampled decoding, and looping is the exact failure mode this logbook exists to track — a greedy log would make the model look worse than it is and blunt the signal. Default `--seed 42`, overridable to deliberately re-roll.

## 2. It would OOM the live training run

```python
os.environ.setdefault("XLA_PYTHON_CLIENT_MEM_FRACTION", "0.5")
```

~3GB of a 6GB card. The #157 run is currently holding 5.2GB. Running this tool as it stands, today, kills the run.

Device must be explicit and recorded: `--device {gpu,cpu}`, GPU as the canonical series, CPU as the fallback that does not require pausing training. They are **not interchangeable** — GPU runs f16, CPU runs f32, so the same weights and the same seed diverge a few tokens in. Device belongs in the filename so the two series can never be silently mixed on a plot.

## 3. It records almost nothing

Written today: step (in the filename), timestamp, temperature, max tokens. Missing: token count, train/val CE, the commit that trained the weights, the commit generating the transcript, device, throughput.

The two commits are distinct and both matter. `run_metadata.json` for #157 records `git_commit: 5455e4d` with `git_dirty: true`; HEAD is now several PRs ahead of that (#198, #200, #202). One SHA cannot tell you which is which.

## The evidence this is worth having

`auxmd.md` (untracked, local) holds a hand-saved inference session from **2026-08-15 17:17**, taken during the gap between run sections 1 and 2 — checkpoint ~step 5,056, ~663M tokens. Against today at step 11,395, 1.494B tokens:

| | Aug 15 | 2026-08-18 | delta |
|---|---|---|---|
| tokens | ~663M | 1.494B | 2.25x |
| train CE | 3.364 | 2.703 | -0.66 |
| val CE | 4.065 | 3.928 | **-0.14** |

Aug 15, `what is an LLM` returned `-M-M-M-M-M-M-M` to the token cap. `the capital of france is` returned *"the ratio of the ratio of the ratio of"* to the token cap. Today, `Photosynthesis is the process by which plants` returns *"use sunlight to store energy"* and a coherent paragraph.

**That entire qualitative gap cost 0.14 nats of held-out CE.** The loss curve massively under-sells it. That is the argument for the logbook: at this stage the number and the behaviour are only loosely coupled, and only one of them is being recorded.

It is also the argument for keeping the logbook **out** of any decision procedure — see non-goals.

## Why this is time-sensitive

Rolling retention is `max_to_keep=3` at 64-step intervals, so a checkpoint exists for roughly 90 minutes and is then deleted. Observed today, three hours apart:

```
20:40   1441791   1449983   1458175   best
23:08   1482751   1490943   1499135   best
```

`1458175` is opt step 11,392 — the checkpoint the samples above came from. It no longer exists.

So the series **cannot be backfilled**. Aug 15's weights are gone (not in cold storage; the collapse rescue only caught 11,007 and 10,689). Every entry has to be captured while its checkpoint is alive. #157 is at 1.494B of 4B; landing this soon buys ~8-10 entries across the back half, landing it late buys the ending and nothing else.

This overlaps #187 (milestone checkpoint tier). If #187 lands, backfilling becomes possible for milestones going forward — but it does not recover anything already evicted, so the two are independent.

## Scope

**Prompt set — frozen, versioned, 8 prompts (`PROMPT_SET_VERSION = 1`).** Changing or removing a prompt bumps the version; appending does not, since each file records which prompts ran.

| # | prompt | probes |
|---|---|---|
| 1 | `the capital of france is` | factual recall, one right answer |
| 2 | `the american flag is red, white and` | loop detector — enumeration invites repetition |
| 3 | `know the truth and the truth will set you` | memorized common text |
| 4 | `The Roman Empire was one of the largest` | multi-sentence explanation |
| 5 | `def fibonacci(n):` | code — 40% of the current curriculum |
| 6 | `2 + 2 = 4. 3 + 5 =` | arithmetic — 25% of the current curriculum |
| 7 | `To make bread, you first need to` | ordered procedure |
| 8 | `Once upon a time, in a small village,` | long-range coherence, no factual anchor |

Prompts 1-4 are the user's, 3 of them carried from `auxmd.md` so the Aug 15 session can be transcribed into the series by hand as a reconstructed entry. Extra prompts passed on the command line are recorded but marked non-standard, so they cannot pollute a comparison.

**Output — markdown with YAML frontmatter.** Metadata parses in three lines; completions stay literal rather than becoming `\n`-escaped JSON strings. Both consumers are served, and the primary consumer is Claude reading it in a future session.

```yaml
---
prompt_set_version: 1
step: 11395
tokens: 1494056960
train_ce: 2.703
val_ce: 3.928
device: gpu
seed: 42
temperature: 0.5
depths: [1, 2, 4, 8]
val_ce_depth: 4
max_steps_limit: 8
tokens_per_second: 42.1
model_commit: 5455e4d
model_commit_dirty: true
tool_commit: e5d00e9
model_arch: refiner
generated: 2026-08-18T23:08:00-03:00
---
```

CE columns come from `metrics.csv` at that step, omitted cleanly when absent (a finished or foreign model has no metrics.csv beside it). `tokens_per_second` excludes the first token, which carries JIT compile and is ~50x slower than steady state.

**Depth — every prompt at every depth, by default.** Depth is not a setting this tool picks; it is an axis it sweeps. Every prompt runs at **depths 1, 2, 4 and 8** unless `--depths` says otherwise, and every completion records the depth that produced it.

Three depths are live in the project simultaneously — training samples randomly up to `MAX_STEPS_LIMIT = 8` (observed average ~4.4 on #157), `VAL_FIXED_DEPTH = 4`, `INFERENCE_DEPTH = 6` — so a transcript that does not name its depth is not interpretable at all. A completion at depth 1 and one at depth 8 are eight-fold different amounts of compute through the same shared block.

The ladder doubles (1 → 2 → 4 → 8) to span the trained range at four points: depth 1 is the degenerate no-refinement baseline, and 8 is `MAX_STEPS_LIMIT`. **4 and 8 are the two that matter most.** Note the default ladder skips `INFERENCE_DEPTH = 6`, which is what every other inference and eval path serves at — `--depths 1,2,4,6,8` is worth considering for that reason, but since the depths that ran are recorded per file, adding 6 later does not invalidate earlier entries.

Frontmatter therefore carries a list, and needs a **separate** field for the validation depth: `val_ce` is measured at depth 4 by the training loop, while completions span the whole ladder. A single `depth:` key would silently claim they describe the same configuration.

**The open question this is built to watch.** Two claims are on the table and neither is settled on *this* model:

1. `trm/config.py` records that *"pretraining shifts the curve LEFT — the same ceiling in fewer loops"*. If true, the depth at which output stops improving should **fall** as #157 goes from 1.5B to 4B tokens.
2. Observed by hand on an earlier, far less trained checkpoint: depth 8 was often best but sometimes *overthought* and came out worse, which is why depth 4 looked like the safer serving default. The dense sweep (`docs/findings/2026-06-19-plan-a-depth-dense-sweep.md`) does not contradict this — it put d6–d8 inside seed noise on a toy-task accuracy metric, which is a different measurement from generated-text quality on the real model.

The **repetition score is measured per depth**, which turns claim 2 from an impression into a number: if depth 8 overthinks, it should loop more than depth 4, and that will show up as a higher 4-gram repetition rate in the same file.

Neither claim gets **settled** here — see non-goals. A hand-read of 8 prompts at one seed is exactly the single-sample comparison the working agreement warns about, with no noise floor and no matched control. What the logbook does is make the effect *visible and dated*, which is what turns "I think 8 overthinks" into something worth pre-registering as a real ablation later.

**Cost.** 8 prompts x 4 depths = 32 completions per snapshot. Trivial on GPU. On the CPU fallback it is the dominant cost of the whole tool, and `--depths 4,8` is the intended escape hatch there.

**Location.** `runs/<run>/transcripts/step_<n>_<device>.md` by default. `runs/` is gitignored, so keepers are promoted by hand into `docs/registry/transcripts/` beside the model cards — matching the registry's curation rule rather than committing every snapshot.

**A repetition number.** Fraction of generated 4-grams that already appeared earlier in the same completion. One deterministic float, **recorded per prompt per depth**. The whole visible difference between Aug 15 and today is looping, and right now that is a vibe rather than a column. Aug 15's flag completion scores near 1.0; today's Roman paragraph is far lower. Cheap, and it makes the thing the eye sees plottable.

## Non-goals

This is a **logbook, not a referee**. Generated text cannot clear a noise floor and cannot attribute cause, so it must never decide KEEP/KILL — that stays with `instruments/verdict.py` and the yardstick. Its jobs are the model card's "why it is notable" line, and the eventual champion-vs-challenger comparison where the difference should be visible and not merely tabulated. Stating this in the docstring is in scope, so nobody in six months mistakes it for evidence.

## Lane

Fully buildable and testable on CPU; CPU is the merge-blocking path. The GPU path is a flag and gets exercised when the card frees after #157 — it does not park this as a draft, since the tool's purpose is served on CPU and holding a finished instrument for 7 days of serial queue would cost more entries than it saves.

**Comment by MatthewLacerda2** [comment]
Added a **Depth** subsection to Scope — it was missing from the original spec.

Depth is a comparability class exactly like device, and three depths are live at once: training samples randomly up to `MAX_STEPS_LIMIT = 8` (avg ~4.4 observed on #157), `VAL_FIXED_DEPTH = 4`, `INFERENCE_DEPTH = 6`.

The specific defect in the first draft: frontmatter carried `val_ce` and the completions under a single implied depth, but `val_ce` is measured at depth 4 by the training loop while generation would run at depth 6. Now two fields — `generation_depth` and `val_ce_depth` — so one file cannot claim they describe the same configuration.

Also added an optional `--depth-strip` (off by default), scoped as an illustration of the settled dense-sweep result rather than a re-measurement of it.

**Comment by MatthewLacerda2** [comment]
Depth section rewritten — it is a **sweep axis, not a setting**.

Every prompt now runs at **depths 1, 2, 4, 8 by default** (`--depths` to override), 32 completions per snapshot. The doubling ladder spans the trained range: 1 is the degenerate no-refinement baseline, 8 is `MAX_STEPS_LIMIT`. 4 and 8 are the two that matter most.

Note the default ladder **skips `INFERENCE_DEPTH = 6`**, which is what every other inference and eval path serves at. Since the depths that ran are recorded per file, adding 6 later does not invalidate earlier entries — but `--depths 1,2,4,6,8` is worth considering.

Frontmatter now carries `depths: [1, 2, 4, 8]` plus a separate `val_ce_depth: 4`, because `val_ce` comes from the training loop at a fixed depth while completions span the ladder.

Added the two open depth claims this is built to watch — the config's "pretraining shifts the curve LEFT" prediction, and the hand-observed "depth 8 sometimes overthinks" — and tied the **repetition score to per-depth recording**, since a depth that overthinks should loop more and that is measurable rather than impressionistic. Neither claim gets settled here; non-goals unchanged.
